L1Trigger/L1THGCal/src/backend_emulator/HGCalHistoSeeding_SA.cc

```diff
@@ -0,0 +1,232 @@
+  for (unsigned int iRow = 0; iRow < config_.cRows(); ++iRow) {
+    for (unsigned int iColumn = 0; iColumn < config_.cColumns(); ++iColumn) {
+      histogramOut.push_back(make_unique<HGCalHistogramCell>(latency, iColumn, iRow));
+    }
+  }
+
+  for (const auto& hc : histogramCells) {
+    const unsigned int binIndex = config_.cColumns() * hc->sortKey() + hc->index();
+    *histogramOut.at(binIndex) += *hc;
+  }
+}
+
+// Smearing at constant r/z using exponentially falling 2^-n kernel
+// Maximum extent of smearing (how many neighbouring bins included) can vary for each r/z bin, and is specified in the configuration
+void HGCalHistoSeeding::smearHistogram1D(HGCalHistogramCellSAPtrCollection& histogram) const {
+  HGCalHistogramCellSACollection lHistogram;
+  for (unsigned int iBin = 0; iBin < histogram.size(); ++iBin) {
+    lHistogram.emplace_back(*histogram.at(iBin));
+  }
+
+  const unsigned int stepLatency = config_.getStepLatency(Step::Smearing1D);
+  for (unsigned int iBin = 0; iBin < lHistogram.size(); ++iBin) {
+    auto& hc = histogram.at(iBin);
+    hc->addLatency(stepLatency);
+
+    const unsigned int col = hc->index();
+    const unsigned int row = hc->sortKey();
+    const unsigned int binIndex = config_.cColumns() * row + col;
+    unsigned int scale = 1;
+    int width = config_.kernelWidth(row);
+    unsigned int offset = 1;
+    while (width > 0) {
+      unique_ptr<HGCalHistogramCell> l1 = make_unique<HGCalHistogramCell>(HGCalHistogramCell());
+      unique_ptr<HGCalHistogramCell> l2 = make_unique<HGCalHistogramCell>(HGCalHistogramCell());
+      unique_ptr<HGCalHistogramCell> r1 = make_unique<HGCalHistogramCell>(HGCalHistogramCell());
+      unique_ptr<HGCalHistogramCell> r2 = make_unique<HGCalHistogramCell>(HGCalHistogramCell());
+
+      if (width >= 2) {
+        if (int(col - offset - 1) >= 0) {
+          l2 = make_unique<HGCalHistogramCell>(lHistogram[binIndex - offset - 1] / 4);
+        }
+        if (int(col + offset + 1) <= int(config_.cColumns() - 1)) {
+          r2 = make_unique<HGCalHistogramCell>(lHistogram[binIndex + offset + 1] / 4);
+        }
+      }
+
+      if (int(col - offset) >= 0) {
+        l1 = make_unique<HGCalHistogramCell>(lHistogram[binIndex - offset] / 2);
+      }
+      if (int(col + offset) <= int(config_.cColumns() - 1)) {
+        r1 = make_unique<HGCalHistogramCell>(lHistogram[binIndex + offset] / 2);
+      }
+      *hc += ((*l2 + *l1) / scale + (*r2 + *r1) / scale);
+      scale *= 4;
+      width -= 2;
+      offset += 2;
+    }
+  }
+}
+
+void HGCalHistoSeeding::normalizeArea(HGCalHistogramCellSAPtrCollection& histogram) const {
+  const unsigned int stepLatency = config_.getStepLatency(NormArea);
+  for (unsigned int iBin = 0; iBin < histogram.size(); ++iBin) {
+    HGCalHistogramCell& hc = *histogram.at(iBin);
+    hc.addLatency(stepLatency);
+    hc.setS(int(1.0 * hc.S() * config_.areaNormalization(hc.sortKey()) / pow(2, config_.nBitsAreaNormLUT() + 1)));
```

Comment:
- Multiplication by `1.0` to be replaced by explicit conversion to float?
- `pow(2, config_.nBitsAreaNormLUT()` to be replaced by `0x1<<config_.nBitsAreaNormLUT()`, more efficient to compute

Reply:
Done both